### PR TITLE
feat: added testFlags option

### PR DESCRIPTION
**Pull request by johan-lejdung**
I tested it out locally and it worked fine and all tests except rungodoc passed but it looked like I had to do some setup for that to pass to that is most likely the reason.

### CODE REVIEW

**Comment by johan-lejdung** [comment]
Solves: https://github.com/Microsoft/vscode-go/issues/1842

File: src/testUtils.ts [hunk 1]
```diff
@@ -166,7 +166,7 @@ export function goTest(testconfig: TestConfig): Thenable<boolean> {
 			outputChannel.show(true);
 		}
 
-		let buildTags: string = testconfig.goConfig['buildTags'];
+		let buildTags: string = testconfig.goConfig['testTags'];
```
Comment (ramya-rao-a) [inline src/testUtils.ts]: I believe we should fallback to using `buildTags` if no `testTags` are provided to ensure we are not breaking anybody else's use case.

File: package.json [hunk 2]
```diff
@@ -515,9 +515,15 @@
         "go.buildTags": {
           "type": "string",
           "default": "",
-          "description": "The Go build tags to use for all commands that support a `-tags '...'` argument",
+          "description": "The Go build tags to use for all commands, excluding `go test` (see testTags), that support a `-tags '...'` argument",
```
Comment (ramya-rao-a) [inline package.json]: `buildtags` will still be used for compiling the test files which will be using `go test`
So we should probably change this to `excluding when tests are run` or something to that effect

File: src/testUtils.ts (hunk 1 above)
Comment (johan-lejdung) [reply src/testUtils.ts]: Of course!

File: src/testUtils.ts [hunk 3]
```diff
@@ -166,7 +166,7 @@ export function goTest(testconfig: TestConfig): Thenable<boolean> {
 			outputChannel.show(true);
 		}
 
-		let buildTags: string = testconfig.goConfig['buildTags'];
+		let testTags: string = (testconfig.goConfig.has("testTags")) ? testconfig.goConfig['testTags'] : testconfig.goConfig['buildTags'];
```
Comment (ramya-rao-a) [inline src/testUtils.ts]: We need to use single quotes, else the linter breaks the build.

File: package.json [hunk 4]
```diff
@@ -515,9 +515,15 @@
         "go.buildTags": {
           "type": "string",
           "default": "",
-          "description": "The Go build tags to use for all commands that support a `-tags '...'` argument",
+          "description": "The Go build tags to use for all commands, excluding when running tests (see testTags), that support a `-tags '...'` argument",
```
Comment (ramya-rao-a) [inline package.json]: The description now is a little confusing. Can we move the exclude statement to the end? How about the below?

`The Go build tags to use for all commands that support a `-tags '...'` argument. When running tests, `go.testTags` will be used instead if it was set.`

File: src/testUtils.ts (hunk 3 above)
Comment (johan-lejdung) [reply src/testUtils.ts]: oops!

File: package.json (hunk 4 above)
Comment (johan-lejdung) [reply package.json]: That does sound better :)

**Comment by ramya-rao-a** [comment]
The `testTags` are to be used when running the tests. The unit tests are using the `check` function that is used to only compile the files and not run tests. Therefore, the `check` method will only use the build tags and will never use the test tags.

That's the reason for the failures

**Comment by johan-lejdung** [comment]
Ah, thanks! Will get it sorted out today. Haven't had a second look yet, sorry for the delay :)

**Comment by johan-lejdung** [comment]
I removed the test for fallback on `buildTags`. The reason was that [this line](https://github.com/Microsoft/vscode-go/pull/1877/commits/c29017cac6cc0edc830cd4c6bba588032d703108#diff-869d156f8d53e34e379625eb2aab2d2cR169) would evaluate to true even though 'testTags' was equal to `''`.

First I chose to modify the check to 
```
let testTags: string = testconfig.goConfig['testTags'].length > 1 ? testconfig.goConfig['testTags'] : testconfig.goConfig['buildTags'];
```

But that would create a weird dependency on `buildTags`, and remove the useCase we are looking to create; where you wanted tests to run without any tags but let it build the files with some tags.

Not sure how to solve the backwards compatibility problem actually since an unset config evaluated to an empty string and not undefined.

**Comment by johan-lejdung** [comment]
Ehm, getting weird errors `Expected 3 arguments, but got 2.`

The function definition is as follows:
```
export function testCurrentFile(goConfig: vscode.WorkspaceConfiguration, args: string[]): Thenable<boolean>
```

Any ideas?

**Comment by ramya-rao-a** [comment]
> Ehm, getting weird errors Expected 3 arguments, but got 2.

Merge from the `master` branch to fix that.

**Comment by ramya-rao-a** [comment]
> and remove the useCase we are looking to create; where you wanted tests to run without any tags but let it build the files with some tags.

We had a similar case with the `go.testFlags` setting. We ended up supported `null` as a valid value. See below:

screen shot 2018-09-02 at 2 50 04 am

**Comment by johan-lejdung** [comment]
That should do it!

**Comment by johan-lejdung** [comment]
When can I expect this to get released? :)

**Comment by johan-lejdung** [comment]
Super, thank you for the help!

**Comment by Rhionin** [comment]
Bless you for building this! I've been dying for this feature!

**Comment by johan-lejdung** [comment]
That's really nice to hear ☺️

**Pull request merged**